docs/language/learn-ql/ql-etudes/river-crossing.ql

```diff
@@ -0,0 +1,114 @@
+    this = "Wolf"
+  }
+}
+
+/** One of two shores. */
+class Shore extends string {
+  Shore() {
+    this = "Left" or
+    this = "Right"
+  }
+
+  /** Returns "the other shore". */
+  Shore flip() {
+    this = "Left" and result = "Right"
+    or
+    this = "Right" and result = "Left"
+  }
+}
+
+/** Renders the state as a string. */
+string renderState(Shore man, Shore goat, Shore cabbage, Shore wolf) {
+  result = man + "," + goat + "," + cabbage + "," + wolf
+}
+
+/** A record of where everything is. */
+class State extends string {
+  Shore man;
+
+  Shore goat;
+
+  Shore cabbage;
+
+  Shore wolf;
+
+  State() { this = renderState(man, goat, cabbage, wolf) }
+
+  /** Returns the state that is reached after ferrying a particular cargo item. */
+  State ferry(Cargo cargo) {
+    cargo = "Nothing" and result = renderState(man.flip(), goat, cabbage, wolf)
+    or
+    cargo = "Goat" and result = renderState(man.flip(), goat.flip(), cabbage, wolf)
+    or
+    cargo = "Cabbage" and result = renderState(man.flip(), goat, cabbage.flip(), wolf)
+    or
+    cargo = "Wolf" and result = renderState(man.flip(), goat, cabbage, wolf.flip())
+  }
+
+  /**
+   * Holds if predator and prey are on the same shore and the man
+   * is on the other shore.
+   */
+  predicate eats(Shore predator, Shore prey) { predator = prey and man = predator.flip() }
+
+  /** Holds if nothing gets eaten in this state. */
+  predicate isSafe() { not (eats(goat, cabbage) or eats(wolf, goat)) }
+
+  /** Returns the state that is reached after safely ferrying a cargo item. */
+  State safeFerry(Cargo cargo) { result = this.ferry(cargo) and result.isSafe() }
+
+  /**
+   * Returns all states that are reachable via safe ferrying.
+   * `path` keeps track of how it is achieved.
+   * `visitedStates` keeps track of previously visited states and is used to avoid loops.
+   */
+  State reachesVia(string path, string visitedStates) {
+    // Trivial case: a state is always reachable from itself.
+    this = result and
+    visitedStates = "" and
+    path = ""
+    or
+    // A state is reachable using pathSoFar and then safely ferrying cargo.
+    exists(string pathSoFar, string visitedStatesSoFar, Cargo cargo |
+      result = this.reachesVia(pathSoFar, visitedStatesSoFar).safeFerry(cargo) and
+      // The resulting state has not yet been visited.
+      not exists(int i | i = visitedStatesSoFar.indexOf(result)) and
+      visitedStates = visitedStatesSoFar + "/" + result and
+      path = pathSoFar + "\n Ferry " + cargo
+    )
+  }
+}
+
+/** The initial state, where everything is on the left shore. */
+class InitialState extends State {
+  InitialState() { this = renderState("Left", "Left", "Left", "Left") }
+}
+
+/** The goal state, where everything is on the right shore. */
+class GoalState extends State {
+  GoalState() { this = renderState("Right", "Right", "Right", "Right") }
+}
+
+from string path
+where any(InitialState is).reachesVia(path, _) = any(GoalState gs)
```

Comment:
A nice feature of high-level code is that it often reads like prose. Perhaps change the variable `is` to `s` to avoid it reading like bad grammar.